### PR TITLE
Rework: Camera Refactoring

### PR DESCRIPTION
**Pull request by tobiasjungmann**
## Camera Refactoring
The image workflow from the Feedback tab was highly coupled with the FeedbackActivity. In order to reuse the functionality of adding images and viewing them as thumbnails in a RecyclerView, it is now a individual module with a minimal API. The API offers four functions. This module can easily integrated in any other activity. It will be reused with the Cafeteria Rating feature.

Instruction on how to add the module to other activities can be found in the ```cameraInterface.kt``` .
The functionality is exactly the same as before. Many functions are just moved to a different file, the actual code changes are smaller than it seems :).

Additionally, a deprecation check for the getBitmap() function in imageUtils was added. 

## How to test:
Add and remove images too the feedback Tab; try to break it while canceling the camera/gallery. Ideally use an Device with an SDK < 29 and one with an SDK >= 29 to test the deprecated function replacement.

### CODE REVIEW

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        ) { result ->
+            onNewImageSelected(result)
+        }
+        this.photoLauncher =
+            parent.registerForActivityResult(StartActivityForResult()) {
+                onNewImageTaken()
+            }
+    }
+
+    private fun processPermissionResult(permissions: Map<String, @JvmSuppressWildcards Boolean>) {
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:81]: I dont fully understand this part:
Could you explain, what `@JvmSuppressWildcards` does in this case?

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraInterface.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+     * It will request the necessary permissions. Once the user selected the image, it is
+     * automatically displayed in the thumbnail adapter and can now be queried with getImagePaths().
+     */
+    fun requestNewImage()
+
+    /**
+     * Clears the list of currently stored images.
+     * Images files are also removed from the storage.
+     */
+    fun clearImages()
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraInterface.kt:40]: From a API-Perpective, probably `deleteStoredImages()` is a better communication of the actual thing this is doing

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        imageRecyclerView.layoutManager =
+            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
+
+        imageRecyclerView.adapter = thumbnailsAdapter
+        this.parent = parent
+        initLaunchers()
+    }
+
+    /**
+     * Launchers must be initialized in the onCreate section of the lifecycle.
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:61]: They are instatiated in `init()`?
Is this equivalent?

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    private fun processPermissionResult(permissions: Map<String, @JvmSuppressWildcards Boolean>) {
+        permissions.entries.forEach {
+            if (it.key == Manifest.permission.READ_EXTERNAL_STORAGE && it.value) {
+                openGallery()
+            } else if (it.key == Manifest.permission.CAMERA && it.value) {
+                takePicture()
+            } else {
+                Log.d("CameraUtils", "Permission Denied")
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:88]: Should we log this like this?
 
```suggestion
                Utils.log("Permission to take a photo/choose it from storage denied")
```

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            photoLauncher.launch(takePictureIntent)
+        } catch (e: ActivityNotFoundException) {
+            photoFile?.delete()
+        }
+    }
+
+    private fun openGallery() {
+        val intent = Intent()
+        intent.type = "image/*"
+        intent.action = Intent.ACTION_GET_CONTENT
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:161]:
```suggestion
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "image/*"
```

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    private fun onNewImageSelected(result: ActivityResult) {
+        val filePath = ImageUtils.rescaleBitmap(context, result.data?.data ?: return) ?: return
+        imageElement.add(filePath)
+        thumbnailsAdapter.addImage(filePath)
+    }
+
+    override fun clearImages() {
+        for (path in imageElement) {
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:193]: See discussion in the interface

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            File(path).delete()
+        }
+    }
+
+    override fun getImagePaths(): Array<String> {
+        return imageElement.toTypedArray()
+    }
+
+    private fun showThumbnailRemovedDialog(path: String) {
+        val builder = AlertDialog.Builder(parent)
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:203]: Consider using `de.tum.in.tumcampusapp.utils.ThemedAlertDialogBuilder` instead. AlertDialog may sometimes not have buttons visible

If this is not possible, add a comment why

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
In general, I really like these changes 👍🏻 

I have left soe inline comments

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt (hunk 1 above)
Comment (tobiasjungmann) [reply app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraManager.kt:81]: That's just a relict and I forgot to remove it. It is not needed in this case.

File: app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraInterface.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Opens a dialog to select whether the image should be taken from the camera/gallery.
+     * It will request the necessary permissions. Once the user selected the image, it is
+     * automatically displayed in the thumbnail adapter and can now be queried with getImagePaths().
+     */
+    fun requestNewImage()
+
+    /**
+     * Clears the list of currently stored images.
+     * Images files are also removed from the storage.
```
[the remark this answers is not in the document]
Comment (tobiasjungmann) [reply app/src/main/java/de/tum/in/tumcampusapp/utils/camera/CameraInterface.kt:38]: New image files are created, even for images from the gallery. Deleting them does therefore only delete the copy, not the original.

**Comment by CommanderStorm** [comment]
If you resolve the linting isssues, this will be merged automatically

**Review by CommanderStorm — APPROVED** [review]

**Pull request merged**